### PR TITLE
[FLINK-25385][table-planner] Harden the function serialization in JSON

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Currently, there is still a lot of Java serialization and unresolved TODOs in serializers. This PR hardens the serialization of `RexNode` and esp. function calls. It supports all kinds of functions and modules.

- Anonymous functions are serialized on a best-effort basis.
- Temporary functions are never serialized.
- Permanent functions are versioned (for built-in functions) and serialized if possible.

## Brief change log

- Adhere to FLIP-190
- Remove Java serialization
- Support function versioning
- Support temporary and module functions

## Verifying this change

This change added tests and can be verified as follows: `RexNodeSerdeTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/bridging/BridgingSqlAggFunction.java [hunk 1]
```diff
@@ -117,6 +116,22 @@ public static BridgingSqlAggFunction of(
                 dataTypeFactory, typeFactory, kind, resolvedFunction, typeInference);
     }
 
+    /** Creates an instance of a aggregate function during translation. */
+    public static BridgingSqlAggFunction of(
+            FlinkContext context,
+            FlinkTypeFactory typeFactory,
+            ContextResolvedFunction resolvedFunction) {
+        final DataTypeFactory dataTypeFactory = context.getCatalogManager().getDataTypeFactory();
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/bridging/BridgingSqlAggFunction.java:124]: That's only personal preference, but since we just hava a couple of chained calls, why not inline them in the `of()` call instead of assigning them to vars?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/SymbolUtil.java [hunk 2]
```diff
@@ -556,8 +595,10 @@ private static void addSymbolMapping(
     }
 
     private static void checkCalciteSymbol(Enum<?> calciteSymbol) {
+        final String className = calciteSymbol.getClass().getName();
         checkArgument(
-                calciteSymbol.getClass().getName().startsWith("org.apache.calcite."),
+                className.startsWith("org.apache.calcite.")
+                        || className.startsWith("com.google.common.collect."),
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/SymbolUtil.java:601]: why is that needed?

**Review by matriv — COMMENTED** [review]
Looks good, Left a few comments/questions.
I will check more carefully the Serializer/Deserializer changes

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/utils/CatalogManagerMocks.java [hunk 3]
```diff
@@ -33,7 +35,15 @@
     public static final String DEFAULT_DATABASE = EnvironmentSettings.DEFAULT_BUILTIN_DATABASE;
 
     public static CatalogManager createEmptyCatalogManager() {
-        final CatalogManager catalogManager = preparedCatalogManager().build();
+        return createEmptyCatalogManager(null);
+    }
+
+    public static CatalogManager createEmptyCatalogManager(@Nullable Catalog catalog) {
+        final CatalogManager.Builder builder = preparedCatalogManager();
+        if (catalog != null) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/utils/CatalogManagerMocks.java:43]: The `preparedCatalogManager()` sets already a catalog, if the given one is null we use the default.

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowTableFunctionJsonPlanTest_jsonplan/testIndividualWindowTVFProcessingTime.out [hunk 4]
```diff
@@ -63,19 +55,16 @@
...
+                  "syntax" : "SPECIAL",
+                  "internalName" : "$-$1",
                   "operands" : [ {
                     "kind" : "INPUT_REF",
                     "inputIndex" : 3,
                     "type" : "TIMESTAMP(3)"
                   }, {
                     "kind" : "LITERAL",
-                    "value" : 1000,
+                    "value" : "1000",
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowTableFunctionJsonPlanTest_jsonplan/testIndividualWindowTVFProcessingTime.out:67]: Yes, to be on the safe side with decimals, we should simply use strings everywhere.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/bridging/BridgingSqlAggFunction.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/bridging/BridgingSqlAggFunction.java:124]: Yes, this might be personal preference. Local variable help in breaking a problem into subproblems. I find this makes the code much more readable and no IDE shows you a warning for this. Not everything that can be inlined should be inlined if you can give it a nice variable name.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/SymbolUtil.java (hunk 2 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/SymbolUtil.java:601]: I will add a comment to this. Unfortunately, for some very specific `RexNode` Calcite relies on Guava enums.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerdeTest.java [hunk 5]
```diff
@@ -0,0 +1,831 @@
...
+            return Optional.empty();
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Test functions
+    // --------------------------------------------------------------------------------------------
+
+    /** Serializable function. */
+    @SuppressWarnings("unused")
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerdeTest.java:774]: Here and below, why do we need to suppress `unused`?

Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerdeTest.java:774]: Ah, it's the variables for the methods throwing exceptions, then maybe nit: to move the `@SuppressWarnings` over those methods.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerdeTest.java [hunk 6]
```diff
@@ -0,0 +1,831 @@
...
+
+    private static void registerTemporaryFunction(SerdeContext serdeContext) {
+        serdeContext
+                .getFlinkContext()
+                .getFunctionCatalog()
+                .registerTemporaryCatalogFunction(
+                        UNRESOLVED_FUNCTION_CAT_ID, NON_SER_FUNCTION_DEF_IMPL, false);
+    }
+
+    private JsonNode serialize(
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerdeTest.java:729]: Checking the calls, we are only testing permanent functions, or am I missing something?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerializer.java [hunk 7]
```diff
@@ -237,33 +247,38 @@ private void serialize(
...
-                for (RexLiteral v : (FlatLists.ComparableList<RexLiteral>) value) {
-                    serialize(v, gen, serializerProvider);
-                }
-                gen.writeEndArray();
+                serializeSargValue((Sarg<?>) value, elementTypeName, gen, serializerProvider);
                 break;
             default:
-                // TODO support ARRAY, MAP
-                throw new TableException("Unknown value: " + value + ", type: " + literalTypeName);
+                throw new TableException(
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerializer.java:273]: You replaced the TODO for `array` &  `map` and removed support for `ROW` and `MULTISET`.
Just double checking, don't have those as literals, right?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerializer.java [hunk 8]
```diff
@@ -335,34 +359,185 @@ private void serialize(RexCall call, JsonGenerator gen, SerializerProvider seria
...
+
+        if (!(definition instanceof UserDefinedFunction)
+                || !isClassNameSerializable((UserDefinedFunction) definition)) {
+            throw cannotSerializePermanentCatalogFunction(objectIdentifier);
+        }
+
+        gen.writeStringField(FIELD_NAME_CLASS, definition.getClass().getName());
+    }
+
+    private static TableException legacyException(String summaryName) {
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerializer.java:510]: Do we have tests for those (also below)?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java [hunk 9]
```diff
@@ -220,185 +260,298 @@ private Object toLiteralValue(
...
+        } catch (Throwable t) {
+            if (throwOnError) {
+                throw new TableException(
+                        String.format("Error during lookup of function '%s'.", identifier), t);
             }
+            return Optional.empty();
+        }
+    }
+
+    private static TableException lookupDisabled(FunctionIdentifier identifier) {
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java:512]: Also tests for those?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/ContextResolvedTableJsonDeserializer.java [hunk 10]
```diff
@@ -172,9 +172,10 @@ private boolean isLookupEnabled(CatalogPlanRestore planRestoreOption) {
     static ValidationException missingIdentifier() {
         return new ValidationException(
                 String.format(
-                        "The table cannot be deserialized, as no identifier is present within the JSON, "
-                                + "but lookup is forced by '%s' == '%s'. "
-                                + "Either allow restoring table from the catalog with '%s' == '%s' | '%s' or make sure you don't use anonymous tables when generating the plan.",
+                        "The table cannot be deserialized as no identifier is present in the persisted plan."
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/ContextResolvedTableJsonDeserializer.java:175]: Unless I'm missing something, those cases (of exceptions) are not covered?
Shouldn't we have tests for all those error messages?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerdeTest.java (hunk 6 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerdeTest.java:729]: true, we can simply the calls

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerializer.java (hunk 7 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonSerializer.java:273]: Never used by Flink. The code generator has also no logic for them.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java (hunk 9 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java:512]: tested in `TestRestoreIdentifier.TestRestoreAllEnforced.withConstantCatalogFunction`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/ContextResolvedTableJsonDeserializer.java (hunk 10 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/ContextResolvedTableJsonDeserializer.java:175]: I agree. I will put this on my list for polishing the table serialization.

**Comment by twalthr** [comment]
Thanks @matriv. I hope I addressed all your comments.

**Review by matriv — APPROVED** [review]
LGTM. Thank you!
Please make a note to add those tests for the exception messages soon.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java [hunk 11]
```diff
@@ -18,58 +18,485 @@
...
+    @DisplayName("Test CatalogPlanCompilation == IDENTIFIER")
+    class TestCompileIdentifier {
+
+        private final CatalogPlanCompilation compilation = CatalogPlanCompilation.IDENTIFIER;
+
+        @Nested
+        @DisplayName("and CatalogPlanRestore == IDENTIFIER")
+        class TestRestoreIdentifier {
+
+            private final CatalogPlanRestore restore = CatalogPlanRestore.IDENTIFIER;
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java]: maybe avoid recreating the ctx everytime but just create one here?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java [hunk 12]
```diff
@@ -92,121 +519,129 @@ public void testRexNodeSerde(RexNode rexNode) throws IOException {
...
-                        FACTORY.createSqlType(SqlTypeName.SMALLINT),
-                        rexBuilder.makeInputRef(FACTORY.createSqlType(SqlTypeName.INTEGER), 1)),
-                // $1 in (1, 3, 5)
-                rexBuilder.makeIn(
-                        rexBuilder.makeInputRef(FACTORY.createSqlType(SqlTypeName.INTEGER), 1),
-                        Arrays.asList(
-                                rexBuilder.makeExactLiteral(new BigDecimal(1)),
-                                rexBuilder.makeExactLiteral(new BigDecimal(3)),
-                                rexBuilder.makeExactLiteral(new BigDecimal(5)))),
+                //
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java]: Why these comments?

**Review by slinkydeveloper — COMMENTED** [review]
I left mostly style comments and small details. The ser/de logic looks good

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java [hunk 13]
```diff
@@ -177,8 +177,9 @@ public String toString() {
     // Shared with BuiltInSqlOperator and BuiltInSqlFunction in planner
     // --------------------------------------------------------------------------------------------
 
+    // note that for SQL operators the name can contain spaces and dollar signs
     private static final Pattern INTERNAL_NAME_PATTERN =
-            Pattern.compile("\\$[A-Z0-9_]+\\$[1-9][0-9]*");
+            Pattern.compile("\\$[A-Z0-9_ $]+\\$[1-9][0-9]*");
```
[the remark this answers is not in the document]
Comment (matriv) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java:182]: +1 for this

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/functions/utils/ScalarSqlFunction.scala [hunk 14]
```diff
@@ -42,7 +44,10 @@ import scala.collection.JavaConverters._
   * @param displayName    name to be displayed in operator name
   * @param scalarFunction scalar function to be called
   * @param typeFactory    type factory for converting Flink's between Calcite's types
+  * @deprecated Use [[BuiltInFunctionDefinitions]] that translates to [[BridgingSqlFunction]].
   */
+@Deprecated
+@deprecated
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/functions/utils/ScalarSqlFunction.scala:50]: yes, one for Scala and one for Java

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java:182]: I neither want to support tabs nor newlines. Only spaces such as for `IS NULL`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/BuiltInSqlOperator.java [hunk 15]
```diff
@@ -74,4 +74,16 @@ static String toQualifiedName(SqlOperator operator) {
...
+        // supports all various kinds of qualified names
+        // $FUNC$1 => FUNC
+        // $IS NULL$1 => IS NULL
+        // $$CALCITE_INTERNAL$1 => $CALCITE_INTERNAL
+        int versionPos = qualifiedName.length() - 1;
+        while (Character.isDigit(qualifiedName.charAt(versionPos))) {
+            versionPos--;
+        }
+        return qualifiedName.substring(1, versionPos);
+    }
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/BuiltInSqlOperator.java:88]: This logic does not break with UTF-16 strings. both `length()` and `charAt` work with special characters. And special characters should also never occur in built-in functions as the INTERNAL_NAME_PATTERN (see above) only allows `A-Z`.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (matriv) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java:182]: Makes sense.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java [hunk 16]
```diff
@@ -220,185 +260,298 @@ private Object toLiteralValue(
...
-            }
-            if (range.hasUpperBound() || range.hasLowerBound()) {
-                builder.add(range);
-            }
-        }
-        boolean containsNull = jsonNode.get(FIELD_NAME_CONTAINS_NULL).booleanValue();
-        return Sarg.of(containsNull, builder.build());
+        final String fieldName = jsonNode.required(FIELD_NAME_NAME).asText();
+        final JsonNode exprNode = jsonNode.required(FIELD_NAME_EXPR);
+        final RexNode refExpr = deserialize(exprNode, serdeContext);
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java:278]: Not that I'm aware of. `RexNode` always end with leaf expressions that reference a literal or field. The field might be a local ref which we currently resolve at all locations.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java [hunk 17]
```diff
@@ -106,78 +118,108 @@ public RexNodeJsonDeserializer() {
...
+                builder.add(range);
+            }
+        }
+        final boolean containsNull = sargNode.required(FIELD_NAME_CONTAINS_NULL).booleanValue();
+        return rexBuilder.makeSearchArgumentLiteral(
+                Sarg.of(containsNull, builder.build()), relDataType);
+    }
+
+    private static Object deserializeLiteralValue(
+            JsonNode literalNode, SqlTypeName sqlTypeName, SerdeContext serdeContext) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeJsonDeserializer.java:221]: For symbols we need to access further fields.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java (hunk 11 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java]: the tests modify the ctx via `dropPermanentFunction` or `registerTemporaryFunction`, it is safer to create a new one per test

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java (hunk 12 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java]: that was a mistake and got removed in following commits

**Comment by twalthr** [comment]
Thanks everyone for the feedback. I hope I could address all of it. I will merge this later today if there are no objections.
